server/src/main/java/org/opensearch/index/query/CombinedFieldsQueryBuilder.java

```diff
@@ -0,0 +1,414 @@
+        NAME,
+        a -> new CombinedFieldsQueryBuilder(a[0])
+    );
+
+    static {
+        XCONTENT_PARSER.declareString(ConstructingObjectParser.constructorArg(), QUERY_FIELD);
+
+        XCONTENT_PARSER.declareStringArray((builder, values) -> {
+            Map<String, Float> fieldToWeight = QueryParserHelper.parseFieldsAndWeights(values);
+            builder.fieldToWeight = fieldToWeight;
+        }, FIELDS_FIELD);
+
+        XCONTENT_PARSER.declareString(CombinedFieldsQueryBuilder::operator, Operator::fromString, OPERATOR_FIELD);
+
+        XCONTENT_PARSER.declareFloat(CombinedFieldsQueryBuilder::boost, BOOST_FIELD);
+        XCONTENT_PARSER.declareString(CombinedFieldsQueryBuilder::queryName, NAME_FIELD);
+        XCONTENT_PARSER.declareField(
+            CombinedFieldsQueryBuilder::minimumShouldMatch,
+            (p, c) -> p.textOrNull(),
+            MINIMUM_SHOULD_MATCH_FIELD,
+            ObjectParser.ValueType.VALUE
+        );
+    }
+
+    public static CombinedFieldsQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        return XCONTENT_PARSER.parse(parser, null);
+    }
+
+    // ========================
+    // ABSTRACT METHOD IMPLEMENTATIONS
+    // ========================
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    // ========================
+    // CORE QUERY BUILDING LOGIC
+    // ========================
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+
+        Map<String, Float> mappedFields = QueryParserHelper.resolveMappingFields(context, fieldToWeight);
+
+        if (mappedFields.isEmpty()) {
+            return Queries.newUnmappedFieldsQuery(fieldToWeight.keySet());
+        }
+
+        List<MappedFieldType> fieldTypes = extractAndValidateMappedFieldTypes(context, mappedFields);
+        Analyzer sharedAnalyzer = validateAndGetSharedAnalyzer(fieldTypes);
+
+        return buildQuery(mappedFields, sharedAnalyzer);
+    }
+
+    // ========================
+    // HELPER METHODS
+    // ========================
+
+    /**
+     * Builds the list of mapped fields.
+     */
+    private List<MappedFieldType> extractAndValidateMappedFieldTypes(QueryShardContext context, Map<String, Float> mappedFields) {
+        List<MappedFieldType> fields = new ArrayList<>();
+
+        for (Map.Entry<String, Float> entry : mappedFields.entrySet()) {
+            String name = entry.getKey();
+            MappedFieldType fieldType = context.getFieldType(name);
+            if (fieldType == null) {
+                continue;
+            }
+
+            validateFieldType(fieldType);
+            fields.add(fieldType);
+        }
+
+        return fields;
+    }
+
+    /**
+     * Validates that the field type is supported by combined fields queries.
+     */
+    private void validateFieldType(MappedFieldType fieldType) {
```

Comment:
Good to see this validation added, coming from this comment in past https://github.com/opensearch-project/OpenSearch/issues/3996#issuecomment-2368877863 seen `combined_field` query used with numeric which is not possible.